flow/display_list_utils.h

```diff
@@ -314,80 +343,233 @@ class DisplayListBoundsCalculator final
    public:
-    SaveLayerWithPaintInfo(DisplayListBoundsCalculator* calculator,
-                           BoundsAccumulator* accumulator,
-                           const SkMatrix& save_matrix,
-                           const SkRect* bounds,
-                           const SkPaint& save_paint);
-    virtual ~SaveLayerWithPaintInfo() = default;
+    RootLayerData() : AccumulatorLayerData(nullptr) {}
+    ~RootLayerData() = default;
 
-    BoundsAccumulator* restore() override;
+   private:
+    FML_DISALLOW_COPY_AND_ASSIGN(RootLayerData);
+  };
 
-   protected:
-    DisplayListBoundsCalculator* calculator_;
+  // Used for |save|
+  class SaveData : public LayerData {
+   public:
+    using LayerData::LayerData;
+    ~SaveData() = default;
+
+   private:
+    FML_DISALLOW_COPY_AND_ASSIGN(SaveData);
+  };
 
-    std::optional<SkRect> bounds_;
-    SkPaint paint_;
+  // Used for |saveLayer|
+  class SaveLayerData : public AccumulatorLayerData {
+   public:
+    SaveLayerData(BoundsAccumulator* outer,
+                  sk_sp<SkImageFilter> filter,
+                  bool paint_nops_on_transparency)
+        : AccumulatorLayerData(outer), layer_filter_(std::move(filter)) {
+      if (!paint_nops_on_transparency) {
+        set_flooded();
+      }
+    }
+    ~SaveLayerData() = default;
+
+    SkRect getLayerBounds() override {
+      SkRect bounds = AccumulatorLayerData::getLayerBounds();
+      if (!getFilteredBounds(bounds, layer_filter_.get())) {
+        set_flooded();
+      }
+      return bounds;
+    }
 
    private:
-    static constexpr SkRect kMissingBounds = SkRect::MakeWH(-1, -1);
+    sk_sp<SkImageFilter> layer_filter_;
 
-    FML_DISALLOW_COPY_AND_ASSIGN(SaveLayerWithPaintInfo);
+    FML_DISALLOW_COPY_AND_ASSIGN(SaveLayerData);
   };
 
-  std::vector<std::unique_ptr<SaveInfo>> saved_infos_;
+  std::vector<std::unique_ptr<LayerData>> layer_infos_;
+
+  // A drawing operation that is not geometric in nature (but which
+  // may still apply a MaskFilter - see |kApplyMaskFilter| below).
+  static constexpr int kIsNonGeometric = 0x00;
+
+  // A geometric operation that is defined as a fill operation
+  // regardless of what the current paint Style is set to.
+  // This flag will automatically assume |kApplyMaskFilter|.
+  static constexpr int kIsFilledGeometry = 0x01;
+
+  // A geometric operation that is defined as a stroke operation
+  // regardless of what the current paint Style is set to.
+  // This flag will automatically assume |kApplyMaskFilter|.
+  static constexpr int kIsStrokedGeometry = 0x02;
+
+  // A geometric operation that may be a stroke or fill operation
+  // depending on the current state of the paint Style attribute.
+  // This flag will automatically assume |kApplyMaskFilter|.
+  static constexpr int kIsDrawnGeometry = 0x04;
+
+  static constexpr int kIsAnyGeometryMask =  //
+      kIsFilledGeometry |                    //
+      kIsStrokedGeometry |                   //
+      kIsDrawnGeometry;
```

Comment:
Consider an `enum class`.

Reply:
(for this and below)

Reply:
These are mask constants...?

Reply:
Yes, you can still define them as an `enum class` to get some more type safety, but I believe you'd have to overload any operators you use to do so. Your call.

Reply:
I like the idea of using an enum if we don't want to make it an `enum class` which would need overloading the operators. Just groups the various values better IMO.

```
enum FlagValue {
  kSomething        = 1 << 0,
  kSomethingElse = 1 << 1,
  kAnotherThing   = 1 << 2,
};
```

Reply:
I see lots of tutorials on this, but none of them answer the basic question of "why".

enums are enumerable. You can see all of its values right there.

Except if it is flags then you don't see all of its values right there. You can make up other values by combining them. Which of the enumerated values is `kIsFrog | kIsZebra`. I don't see a "zebra frog" in the list and such an animal would not likely survive long.

So, this concept has philosophical problems.

What advantages does it provide? "Hey, it's an enum!" - and how does that help anything?

One of the advantages of enums is knowing that your switch statements are complete, but you don't use flags in switches and if you did this aspect of enums would lie to you. As long as you had a `case` for each bit flag (which would be a horribly inefficient switch table) then it would think things were complete.

I suppose it has a type associated with it, but I feel like it is expressing "apples" as if they were "oranges".

Reply:
If you use an enum class the compiler will help make sure you use the right types and don't end up accidentally mixing things up.  But it does require operator overloading